File: problems/383-ransom-note/solution.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        mag_count = {}
        for char in magazine:
            mag_count[char] = mag_count.get(char, 0) + 1
        
        for char in ransomNote:
            if char not in mag_count or mag_count[char] == 0:
                return False
            mag_count[char] -= 1
        return True

File: problems/383-ransom-note/solution.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        int[] magCount = new int[26];
        int[] ranCount = new int[26];

        for (char c : magazine.toCharArray()) {
            magCount[c - 'a']++;
        }

        for (char c : ransomNote.toCharArray()) {
            ranCount[c - 'a']++;
        }

        for (int i = 0; i < 26; i++) {
            if (ranCount[i] > magCount[i]) {
                return false;
            }
        }

        return true;
    }
}